runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/streaming/ActiveWorkState.java

```diff
@@ -120,6 +123,42 @@ synchronized ActivateWorkResult activateWorkForKey(ShardedKey shardedKey, Work w
     return ActivateWorkResult.QUEUED;
   }
 
+  public static class FailedTokens {
+    public long workToken;
+    public long cacheToken;
+
+    public FailedTokens(long workToken, long cacheToken) {
+      this.workToken = workToken;
+      this.cacheToken = cacheToken;
+    }
+  }
+
+  synchronized void failWorkForKey(Map<Long, List<FailedTokens>> failedWork) {
+    // Note we can't construct a ShardedKey and look it up in activeWork directly since
```

Comment:
we could change activeMap to a multimap keyed by sharding_key.
Lookup would have to find all entries with same sharding_key and compare user key, but we don't expect collisions.

could also keep map keyed by sharding_key and use the existing queue.  We'd want to change the queueing logic to support possibly different keys though.  But I don't think serializing procesisng in the case of such a collision is a real problem.

Reply:
We spoke offline. We're going to do this in a separate PR.